File: 3_src/1_data/fetch_sinmod_data.R
library(ncdf4)

fetch_sinmod_data <- function(sinmod_data, sampled_locations, env_variables) {
    # Get no_sites # nolint
    no_sites <- nrow(sampled_locations)

    # Create an empty data frame to store the SINMOD data for multiple variables
    sinmod_df <- data.frame(matrix(NA, nrow = no_sites, ncol = length(env_variables)))
    colnames(sinmod_df) <- env_variables # Assign variable names as column names

    # Fetch data for each variable one by one at the sampled locations (avoid loading entire dataset)
    for (j in 1:length(env_variables)) {
        var <- env_variables[j]

        # Get dimensions of the variable (3D or 4D)
        var_dims <- length(sinmod_data$var[['temperature']]$dim)

        if (var_dims == 3) {
            # Fetch data for 3D variables like "elevation"
            for (i in 1:no_sites) {
                x <- sampled_locations[i, 1] # x-coordinate
                y <- sampled_locations[i, 2] # y-coordinate
                # Fetch data at the specific location
                sinmod_df[i, j] <- ncvar_get(sinmod_data, var, start = c(x, y, 1), count = c(1, 1, 1))
            }
        } else {
            # Fetch data for 4D variables like "temperature"
            for (i in 1:no_sites) {
                x <- sampled_locations[i, 1] # x-coordinate
                y <- sampled_locations[i, 2] # y-coordinate
                # Fetch data at the specific location and time/depth slice (time=0, depth=0)
                sinmod_df[i, j] <- ncvar_get(sinmod_data, var, start = c(x, y, 1, 1), count = c(1, 1, 1, 1))
            }
        }
    }
    return(sinmod_df)
}

File: 3_src/1_data/fetch_sinmod_data.py
import xarray as xr
import pandas as pd

def fetch_sinmod_data(sinmod_data_path, sampled_locations, env_variables, target=True, normalise=False):
    """
    Fetch SINMOD data for specific (x, y, time) combinations defined in sampled_locations.

    Parameters:
        sinmod_data_path (str): Path to the SINMOD NetCDF file.
        sampled_locations (pd.DataFrame): DataFrame with columns ['x', 'y', 'time'] for sample locations.
        env_variables (list): List of environmental variable names to extract.

    Returns:
        pd.DataFrame: DataFrame with SINMOD data for the specified variables and locations.
    """
    # Open the SINMOD NetCDF file
    sinmod_data = xr.open_dataset(sinmod_data_path)
    
    # Prepare the result DataFrame by copying sampled_locations
    result = sampled_locations.copy()

    # Iterate over each environmental variable
    for var in env_variables:
        if var not in sinmod_data.variables:
            raise ValueError(f"Variable '{var}' not found in the SINMOD dataset.")
        
        if target:
            # Fetch data for each row in sampled_locations
            result[var] = sampled_locations.apply(
                lambda row: sinmod_data[var].isel(time=int(row['time_idx'])).sel(
                    xc=row['x'], 
                    yc=row['y'], 
                    method="nearest"
                ).values.item(),
                axis=1
            )
        else:
            # Iterate over each `stat` dimension for the variable
            for stat in sinmod_data[var].stat.values:
                # Fetch data for each stat dimension and label column as `{var}_{stat}`
                result[f"{var}_{stat}"] = sampled_locations.apply(
                    lambda row: sinmod_data[var].sel(stat=stat).sel(xc=row['x'], yc=row['y'], method="nearest").values.item(),
                    axis=1
                )
    # Close the dataset
    sinmod_data.close()

    if target:
        result[env_variables] = result[env_variables].fillna(0)
        if normalise:
            result[env_variables] = result[env_variables].div(result[env_variables].sum(axis=1), axis=0)
            
    return result
